docs/design_notes/lift.md

```diff
@@ -0,0 +1,355 @@
+    If a collection or PRNG sequence does not match any filter it will not be lifted.
+    This means that it cannot be used inside the transformation and and attempting to do this will cause an error to be raised.
+    For example, `in_vars = (["params"], True)` will cause the "params" collection to be put in the first group and all other collection to be put in the second group.
+
+    For each PRNG sequence that is matched we seed a new PRNG sequence by calling `make_rng`.
+    This avoids the need to update the PRNG state after the lifted transformation is complete.
+
+3. *Transform-specific lifting*
+
+    `fn` is called with the variable and PRNG groups.
+    JAX transforms have varying signatures and lifting options. Arguably the cleanest example is `vmap`.
+    In the case of vmap the function arguments, PRNGs and variable collections are passed into a `jax.vmap` wrapped function.
+
+4. *Scope reconstruction*
+
+    Now that the variables and PRNGs are lifted inside the transformation, we want to recreate the lifted scopes. Pack calls
+    `fn` with a `scope_fn` that takes the lifted variables and PRNGs and returns the reconstructed scopes with the lifted variables and rng sequences.
+
+5. *Repack stage*
+
+    After we have used the lifted scopes we have to retrieve the updated variables (PRNG sequences can simply be discarded).
+    pack passes the `repack_fn` to support this.
+    This stage is similar to stage 2 except that we only lift variables and immutable variables are ignored.
+    Immutable variables cannot be updated. Therefore, they should not be returned from the transformed function. 
+
+6. *Commit stage*
+
+    `pack` expects `fn` to return a pair where the first item will simply be returned from pack and the second item should be the repacked variables.
+    The updated variables are stored in the original/un-lifted scopes such that the mutations that happen inside the transformation survive after the transformation completes.
+
+
+### Using pack example
+
+
+A minimal example of using `pack` to transpose each matrix in a variable collection:
+
+```python
+from flax.core import lift
+from flax.core import Scope, init, apply, nn as core_nn
+
+def lift_transpose(fn, target='params', variables=True, rngs=True)
+  # by default we transpose 'params' and simply pass through all other variables.
+  def wrapper(scope_fn, repack_fn, variable_groups, rng_groups, *args):
+    # normally we would first call into a JAX transformed function here...
+    target, rest = variable_groups
+    def trans(x):
+      if x.ndim == 2:
+        return x.T
+      return x
+    target = jax.tree_map(trans, target)
+    variable_groups = (target, rest)
+    scope = scope_fn(variable_groups, rng_groups)
+    y = fn(scope, *args)
+    out_variables = repack_fn(scope)
+    return y, out_variables
+  return lift.pack(wrapper, (target, variables), (variables,), (rngs,))
```

Comment:
I would suggest a comment here with the signature of `lift.pack(fn, in_vars, out_vars, rngs)`